src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java

```diff
@@ -341,28 +344,38 @@ boolean canAccessRepository(Long participationId) throws IllegalArgumentExceptio
             return forbidden();
         }
 
-        Optional<Submission> optionalSubmission = participation.getSubmissions().stream().findFirst();
-        if (optionalSubmission.isEmpty()) {
+        ProgrammingSubmission programmingSubmission = (ProgrammingSubmission) participation.getSubmissions().stream().findFirst().orElse(null);
+        // If a resultId is specified and the ID does not belong to the latest result, find the corresponding submission. Otherwise use the latest submission.
+        if (resultId.isPresent() && (programmingSubmission == null || programmingSubmission.getResults().stream().noneMatch(r -> resultId.get().equals(r.getId())))) {
+            // Note: if the submission was null, this will fail with a good message.
+            programmingSubmission = programmingSubmissionRepository.findByResultIdElseThrow(resultId.get());
+            if (!Objects.equals(participation.getId(), programmingSubmission.getParticipation().getId())) {
+                // The result of the given ID must belong to the participation
+                log.warn("Participation ID {} tried to access the build logs of another participation's submission with ID {}.", participation.getId(),
+                        programmingSubmission.getId());
+                return forbidden();
```

Comment:
I'd think bad request would be more fitting, given the numbers of the request dont match, its not really that a permission was checked

Reply:
+1